daemon/migrator.c

```diff
@@ -0,0 +1,804 @@
+}
+
+/* ------------------------------------------------------------------------- *
+ * APPROVAL_ATTRIBUTES
+ * ------------------------------------------------------------------------- */
+
+static bool
+approval_valid(const approval_t *self)
+{
+    return approval_uid(self) != MIGRATION_UID_UNDEFINED;
+}
+
+static const gchar *
+approval_profile(const approval_t *self)
+{
+    /* Application profile path, i.e. application's desktop entry file path */
+    return self->apr_profile;
+}
+
+static uid_t
+approval_uid(const approval_t *self)
+{
+    /* UID from approval file path */
+    return self->apr_uid;
+}
+
+static stringset_t *
+approval_permissions(const approval_t *self)
+{
+    /* Approved permissions */
+    return self->apr_permissions;
+}
+
+static const gchar *
+approval_organization(const approval_t *self)
+{
+    /* The same as OrganizationName in Sailjail section */
+    return self->apr_organization;
+}
+
+static const gchar *
+approval_application(const approval_t *self)
+{
+    /* The same as ApplicationName in Sailjail section */
+    return self->apr_application;
+}
+
+#if DEAD_CODE
+static const gchar *
+approval_exec(const approval_t *self)
+{
+    /* Binary path, does not contain arguments */
+    return self->apr_exec;
+}
+#endif
+
+static const migrator_t *
+approval_migrator(const approval_t *self)
+{
+    return self->apr_migrator;
+}
+
+static const control_t *
+approval_control(const approval_t *self)
+{
+    return migrator_control(approval_migrator(self));
+}
+
+/* ------------------------------------------------------------------------- *
+ * UTILITY
+ * ------------------------------------------------------------------------- */
+
+#if DEAD_CODE
+static bool
+exec_matches_binary(const gchar *exec, const gchar *binary_path)
+{
+    gchar **argv = NULL;
+    gint    argc = 0;
+    bool    rv   = false;
+
+    /* Note: This can deal only with Exec lines that contain the
+     *       application as the first argument, i.e. it can't deal with
+     *       Exec lines that contains sailjail or invoker first.
+     */
+    if( g_shell_parse_argv(exec, &argc, &argv, NULL) &&
+        argc >= 1 && 
+        !g_strcmp0(argv[0], binary_path) )
+        rv = true;
+
+    g_strfreev(argv);
+    return rv;
+}
+#endif
+
+static void
+remove_approval_path(const gchar *path)
+{
+    /* Remove path and any empty parent directories until HOMESCREEN_DATA_PATH */
+    if( unlink(path) == 0 ) {
+        gchar *dir_path = path_dirname(path);
+        while( may_remove_approval_path(dir_path) && rmdir(dir_path) == 0 ) {
+            gchar *temp = dir_path;
+            dir_path = path_dirname(temp);
+            g_free(temp);
+        }
+        log_debug("%s cleaned up to %s", path, dir_path);
+        g_free(dir_path);
+    }
+    else {
+        log_err("%s could not remove: %m", path);
+    }
+}
+
+static bool
+may_remove_approval_path(const gchar *dir_path)
+{
+    return strlen(dir_path) > strlen(HOMESCREEN_DATA_PATH);
```

Comment:
This looks like accident waiting to happen. Maybe something like ...

    size_t n = strlen(HOMESCREEN_DATA_PATH);
    return !strncmp(dir_path, HOMESCREEN_DATA_PATH,n) && dir_path[n] != 0;

Reply:
Or, since removal/ignoring of HOMESCREEN_DATA_PATH prefix seems to have more use below too, one might have a helper function for that purpose and use it instead of making assumptions about the prefix being there?

Reply:
Added that helper function and rewrote the rest. It should be better now.